addons/crm_mail_plugin/controllers/mail_plugin.py

```diff
@@ -1,81 +1,153 @@
-            recurring_plan: the value of the recurring plan field if the lead has a recurring revenue
-        }
-        """
+        return super()._get_record_redirect_url(model, record_id)
 
-        partner_leads = request.env['crm.lead'].search(
-            [('partner_id', '=', partner.id)], offset=offset, limit=limit)
-        recurring_revenues = request.env.user.has_group('crm.group_use_recurring_revenues')
+    @route()
+    def search_records(self, model, query, limit=30):
+        if model == "crm.lead":
+            if model not in self._mail_models_access_whitelist('read'):
+                raise Forbidden()
 
-        leads = []
-        for lead in partner_leads:
-            lead_values = {
-                'lead_id': lead.id,
-                'name': lead.name,
-                'expected_revenue': formatLang(request.env, lead.expected_revenue,
-                                               currency_obj=lead.company_currency),
-                'probability': lead.probability,
-            }
+            return self._search_leads([('name', 'ilike', query)], limit=limit)
 
-            if recurring_revenues:
-                lead_values.update({
-                    'recurring_revenue': formatLang(request.env, lead.recurring_revenue,
-                                                    currency_obj=lead.company_currency),
-                    'recurring_plan': lead.recurring_plan.name,
-                })
+        return super().search_records(model, query, limit)
 
-            leads.append(lead_values)
+    def _search_leads(self, domain, limit=30):
+        """Search the leads base on the domain and format the result.
 
-        return leads
+        The lead structure is:
+        {
+            id: the lead's id,
+            name: the lead's name,
+            revenues_description: the expected revenue with probability
+        }
+        """
+        lead_count = request.env['crm.lead'].search_count(domain)
+        partner_leads = request.env['crm.lead'].search(domain, limit=limit, order='won_status')
+        leads = [self._format_lead(lead) for lead in partner_leads]
+        return leads, lead_count
 
-    def _get_contact_data(self, partner):
+    def _get_contact_data(self, partner, email, **kwargs):
         """
         Return the leads key only if the current user can create leads. So, if they can not
         create leads, the section won't be visible on the addin side (like if the CRM
         module was not installed on the database).
         """
-        contact_values = super(MailPluginController, self)._get_contact_data(partner)
+        contact_values = super()._get_contact_data(partner, email, **kwargs)
 
         if not request.env['crm.lead'].has_access('create'):
             return contact_values
 
-        if not partner:
+        domain = []
+        if partner:
+            domain.append([('partner_id', '=', partner.id)])
+        if email:
```

Comment:
Can we receive both partner and email ? I see we do an "or", but does that make sense ? If not, probably have a if - elif structure instead. Could also search on partner + partner.email_normalized if we have a partner, then fallback on email for example.

Reply:
> Can we receive both partner and email ? I see we do an "or", but does that make sense ?

Yes, for the case where you have the email of the partner on the lead, but not the partner itself

> Could also search on partner + partner.email_normalized if we have a partner, then fallback on email for example.

Ok to do that (+ the elif structure)